### PR TITLE
[SLP]Fix PR87133: crash because of different altopcodes for cmps after reordering.

### PR DESCRIPTION
**Pull request by alexey-bataev**
If the node has cmp instruction with 3 or more different but swappable
predicates, need to keep same kind of main/alternate opcodes to avoid
incorrect detection of opcodes after reordering. Reordering changes the
order and we may erroneously consider swappable opcodes as
non-compatible/alternate, which may lead to a later compiler crash.

### CODE REVIEW

**Review by wjristow — COMMENTED** [review]
I'm not really qualified to review this, but I can confirm that this change also fixes my original (non-reduced) test-case.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -658,6 +658,32 @@ static InstructionsState getSameOpcode(ArrayRef<Value *> VL,
...
+      return false;
+    SetVector<unsigned> UniquePreds, UniqueNonSwappedPreds;
+    UniquePreds.insert(BasePred);
+    UniqueNonSwappedPreds.insert(BasePred);
+    for (Value *V : VL) {
+      auto *I = dyn_cast<CmpInst>(V);
+      if (!I) {
+        UniquePreds.clear();
+        UniqueNonSwappedPreds.clear();
+        break;
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:672]: why not just return false?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:672]: Yep, remains from the initial version of the patch, will fix.

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by AtariDreams** [comment]
/cherry-pick https://github.com/llvm/llvm-project/commit/cd29126b6333c28cc4df7b932ed0d6d6c13983d1

**Comment by RKSimon** [comment]
@AtariDreams #87133 indicates this has been broken since llvm15 - we don't cherry pick to the current release branch after its first release unless the fix is for a regression from the previous release only (17.x for 18.x etc.).

**Pull request merged**